File: macos/Sources/ChoroConvert/ONNXProgram.swift
import Foundation

/// ONNX のモデルを組み立てる。
///
/// Core ML の `.mlmodel` と同じく protobuf なので、書き手（`Protowire`）はそのまま使える。
/// 変わるのは番号と綴りだけである。
///
/// **番号は仕様から写す。** `onnx/onnx.proto3`（github.com/onnx/onnx）を見て写した。
/// MIL のときと同じ作法で、写し間違えれば ONNX Runtime が読めずにその場で落ちる。
///
/// **Core ML より易しいところが 2 つある。**
///
/// - 長さが動く。バケット（固定長を何本も束ねる）が要らない
/// - 重みはモデルの中に置ける。別の `weight.bin` を持たなくてよい
///
/// 逆に難しいのは、演算の綴りが Core ML と違うことだけである。
public struct ONNXProgram {
    /// protobuf の番号。仕様から写したもの。
    private enum Model {
        static let irVersion = 1, producerName = 2, producerVersion = 3
        static let graph = 7, opsetImport = 8
    }

    private enum OperatorSet {
        static let domain = 1, version = 2
    }

    private enum Graph {
        static let node = 1, name = 2, initializer = 5, input = 11, output = 12
    }

    private enum Node {
        static let input = 1, output = 2, name = 3, opType = 4, attribute = 5
    }

    private enum Attribute {
        static let name = 1, f = 2, i = 3, s = 4, t = 5, floats = 7, ints = 8, type = 20
    }

    private enum Tensor {
        static let dims = 1, dataType = 2, name = 8, rawData = 9
    }

    private enum ValueInfo {
        static let name = 1, type = 2
    }

    private enum TypeProto {
        static let tensorType = 1
    }

    private enum TensorTypeProto {
        static let elemType = 1, shape = 2
    }

    private enum Shape {
        static let dim = 1
    }

    private enum Dimension {
        static let value = 1, param = 2
    }

    /// 値の型。仕様の `TensorProto.DataType` から写した。
    public enum DataType: Int {
        case float = 1
        case int64 = 7
        case float16 = 10
    }

    /// 属性の型。仕様の `AttributeProto.AttributeType` から写した。
    private enum AttributeType: Int {
        case float = 1, int = 2, string = 3, tensor = 4
        case floats = 6, ints = 7
    }

    /// 大きさ。動く軸は名前で置く。
    public enum Extent {
        case fixed(Int)
        /// 長さが決まらない軸。同じ名前は同じ長さを指す。
        case named(String)
    }

    /// 演算 1 つ。
    public struct Op {
        public var type: String
        public var inputs: [String]
        public var outputs: [String]
        public var attributes: [Attr]

        public init(_ type: String, inputs: [String], outputs: [String], attributes: [Attr] = []) {
            self.type = type
            self.inputs = inputs
            self.outputs = outputs
            self.attributes = attributes
        }
    }

    /// 演算に添える値。
    public enum Attr {
        case int(String, Int)
        case ints(String, [Int])
        case float(String, Float)
        case floats(String, [Float])
        case string(String, String)
    }

    /// 重み。名前で参照される定数。
    public struct Initializer {
        public var name: String
        public var dims: [Int]
        public var type: DataType
        /// そのまま並べたバイト列（little endian）。
        public var raw: [UInt8]

        public init(name: String, dims: [Int], type: DataType, raw: [UInt8]) {
            self.name = name
            self.dims = dims
            self.type = type
            self.raw = raw
        }
    }

    /// 出入口 1 つ。
    public struct Port {
        public var name: String
        public var type: DataType
        public var shape: [Extent]

        public init(name: String, type: DataType, shape: [Extent]) {
            self.name = name
            self.type = type
            self.shape = shape
        }
    }

    public var name: String
    public var opset: Int
    public var inputs: [Port] = []
    public var outputs: [Port] = []
    public var initializers: [Initializer] = []
    public var ops: [Op] = []

    /// - Parameter opset: 既定は 17。ModernBERT が要る演算はすべてこの版に入っている。
    public init(name: String, opset: Int = 17) {
        self.name = name
        self.opset = opset
    }

    // MARK: - 組み立て

    /// 読める形にする。
    ///
    /// **IR の版は 8 に置く。** 新しい版を名乗ると、少し古い ONNX Runtime が読まない。
    /// opset 17 に要るのは 8 までなので、名乗る理由が無い。
    public func encoded() -> Data {
        var model = Protowire()
        model.field(Model.irVersion, varint: 8)
        model.field(Model.producerName, string: "choro-convert")
        model.field(Model.producerVersion, string: "1")
        model.field(Model.opsetImport, message: Protowire.message { set in
            // 既定の演算集合は domain を空文字で書く。省くと版が付かない。
            set.field(OperatorSet.domain, string: "")
            set.field(OperatorSet.version, varint: opset)
        })
        model.field(Model.graph, message: graph())
        return model.data
    }

    private func graph() -> Protowire {
        Protowire.message { graph in
            graph.field(Graph.name, string: name)
            for op in ops {
                graph.field(Graph.node, message: node(op))
            }
            for weight in initializers {
                graph.field(Graph.initializer, message: tensor(weight))
            }
            for port in inputs {
                graph.field(Graph.input, message: valueInfo(port))
            }
            for port in outputs {
                graph.field(Graph.output, message: valueInfo(port))
            }
        }
    }

    private func node(_ op: Op) -> Protowire {
        Protowire.message { node in
            // **入と出は宣言した順に並べる。** ONNX は位置で意味が決まる。
            for input in op.inputs {
                node.field(Node.input, string: input)
            }
            for output in op.outputs {
                node.field(Node.output, string: output)
            }
            node.field(Node.name, string: op.outputs.first ?? op.type)
            node.field(Node.opType, string: op.type)
            for attribute in op.attributes {
                node.field(Node.attribute, message: attr(attribute))
            }
        }
    }

    private func attr(_ attribute: Attr) -> Protowire {
        Protowire.message { out in
            switch attribute {
            case let .int(name, value):
                out.field(Attribute.name, string: name)
                out.field(Attribute.type, varint: AttributeType.int.rawValue)
                out.field(Attribute.i, varint: UInt64(bitPattern: Int64(value)))
            case let .ints(name, values):
                out.field(Attribute.name, string: name)
                out.field(Attribute.type, varint: AttributeType.ints.rawValue)
                // packed で書く。ONNX は packed も繰り返しも受ける。
                var packed = [UInt8]()
                for value in values {
                    Protowire.appendVarint(UInt64(bitPattern: Int64(value)), to: &packed)
                }
                out.field(Attribute.ints, bytes: packed)
            case let .float(name, value):
                out.field(Attribute.name, string: name)
                out.field(Attribute.type, varint: AttributeType.float.rawValue)
                out.field(Attribute.f, float: value)
            case let .floats(name, values):
                out.field(Attribute.name, string: name)
                out.field(Attribute.type, varint: AttributeType.floats.rawValue)
                var packed = [UInt8]()
                for value in values {
                    withUnsafeBytes(of: value.bitPattern.littleEndian) { packed.append(contentsOf: $0) }
                }
                out.field(Attribute.floats, bytes: packed)
            case let .string(name, value):
                out.field(Attribute.name, string: name)
                out.field(Attribute.type, varint: AttributeType.string.rawValue)
                out.field(Attribute.s, bytes: Array(value.utf8))
            }
        }
    }

    private func tensor(_ weight: Initializer) -> Protowire {
        Protowire.message { out in
            for size in weight.dims {
                out.field(Tensor.dims, varint: size)
            }
            out.field(Tensor.dataType, varint: weight.type.rawValue)
            out.field(Tensor.name, string: weight.name)
            out.field(Tensor.rawData, bytes: weight.raw)
        }
    }

    private func valueInfo(_ port: Port) -> Protowire {
        Protowire.message { out in
            out.field(ValueInfo.name, string: port.name)
            out.field(ValueInfo.type, message: Protowire.message { type in
                type.field(TypeProto.tensorType, message: Protowire.message { tensor in
                    tensor.field(TensorTypeProto.elemType, varint: port.type.rawValue)
                    tensor.field(TensorTypeProto.shape, message: Protowire.message { shape in
                        for extent in port.shape {
                            shape.field(Shape.dim, message: Protowire.message { dim in
                                switch extent {
                                case let .fixed(size):
                                    dim.field(Dimension.value, varint: size)
                                case let .named(label):
                                    dim.field(Dimension.param, string: label)
                                }
                            })
                        }
                    })
                })
            })
        }
    }
}

File: windows/ChoroReader.Convert/OnnxProgram.cs
using System.Text;

namespace ChoroReader.Convert;

/// <summary>値の型。仕様の <c>TensorProto.DataType</c> から写した。</summary>
public enum OnnxType
{
    Float = 1,
    Int64 = 7,
    Float16 = 10,
}

/// <summary>大きさ。動く軸は名前で置く。</summary>
public readonly record struct Extent(int Fixed, string? Name)
{
    public static Extent Of(int size) => new(size, null);

    /// <summary>長さが決まらない軸。同じ名前は同じ長さを指す。</summary>
    public static Extent Named(string label) => new(0, label);
}

/// <summary>演算に添える値。</summary>
public abstract record Attr(string Name)
{
    public sealed record Int(string Name, long Value) : Attr(Name);
    public sealed record Ints(string Name, IReadOnlyList<long> Values) : Attr(Name);
    public sealed record Float(string Name, float Value) : Attr(Name);
    public sealed record Text(string Name, string Value) : Attr(Name);
}

/// <summary>演算 1 つ。</summary>
public sealed record Op(string Type, IReadOnlyList<string> Inputs, IReadOnlyList<string> Outputs,
                        IReadOnlyList<Attr> Attributes);

/// <summary>重み。名前で参照される定数。</summary>
public sealed record Initializer(string Name, IReadOnlyList<int> Dims, OnnxType Type, byte[] Raw);

/// <summary>出入口 1 つ。</summary>
public sealed record Port(string Name, OnnxType Type, IReadOnlyList<Extent> Shape);

/// <summary>
/// ONNX のモデルを組み立てる。
///
/// <para>
/// <b>番号は仕様から写す。</b><c>onnx/onnx.proto3</c>（github.com/onnx/onnx）を見て写した。
/// 写し間違えれば ONNX Runtime が読めずにその場で落ちる。
/// </para>
/// <para>
/// Core ML より易しいところが 2 つある。長さが動くのでバケットが要らない。
/// 重みをモデルの中に置けるので、別のファイルを持たなくてよい。
/// </para>
/// </summary>
public sealed class OnnxProgram(string name, int opset = 17)
{
    private static class Model
    {
        public const int IrVersion = 1, ProducerName = 2, ProducerVersion = 3, Graph = 7, OpsetImport = 8;
    }

    private static class OperatorSet
    {
        public const int Domain = 1, Version = 2;
    }

    private static class GraphFields
    {
        public const int Node = 1, Name = 2, Initializer = 5, Input = 11, Output = 12;
    }

    private static class NodeFields
    {
        public const int Input = 1, Output = 2, Name = 3, OpType = 4, Attribute = 5;
    }

    private static class AttributeFields
    {
        public const int Name = 1, F = 2, I = 3, S = 4, Floats = 7, Ints = 8, Type = 20;
    }

    private static class TensorFields
    {
        public const int Dims = 1, DataType = 2, Name = 8, RawData = 9;
    }

    private static class ValueInfoFields
    {
        public const int Name = 1, Type = 2;
    }

    private static class TypeFields
    {
        public const int TensorType = 1;
    }

    private static class TensorTypeFields
    {
        public const int ElemType = 1, Shape = 2;
    }

    private static class ShapeFields
    {
        public const int Dim = 1;
    }

    private static class DimensionFields
    {
        public const int Value = 1, Param = 2;
    }

    /// <summary>属性の型。仕様の <c>AttributeProto.AttributeType</c> から写した。</summary>
    private enum AttributeType
    {
        Float = 1, Int = 2, String = 3, Floats = 6, Ints = 7,
    }

    public string Name { get; } = name;
    public int Opset { get; } = opset;
    public List<Port> Inputs { get; } = [];
    public List<Port> Outputs { get; } = [];
    public List<Initializer> Initializers { get; } = [];
    public List<Op> Ops { get; } = [];

    /// <summary>
    /// 読める形にする。
    ///
    /// <para>
    /// <b>IR の版は 8 に置く。</b>新しい版を名乗ると、少し古い ONNX Runtime が読まない。
    /// opset 17 に要るのは 8 までなので、名乗る理由が無い。
    /// </para>
    /// </summary>
    public byte[] Encode()
    {
        var model = new Protowire();
        model.Field(Model.IrVersion, 8ul);
        model.Field(Model.ProducerName, "choro-convert");
        model.Field(Model.ProducerVersion, "1");
        model.Field(Model.OpsetImport, Protowire.Message(set =>
        {
            // 既定の演算集合は domain を空文字で書く。省くと版が付かない。
            set.Field(OperatorSet.Domain, "");
            set.Field(OperatorSet.Version, Opset);
        }));
        model.Field(Model.Graph, Graph());
        return model.Bytes;
    }

    private Protowire Graph() => Protowire.Message(graph =>
    {
        graph.Field(GraphFields.Name, Name);
        foreach (var op in Ops)
        {
            graph.Field(GraphFields.Node, Node(op));
        }
        foreach (var weight in Initializers)
        {
            graph.Field(GraphFields.Initializer, Tensor(weight));
        }
        foreach (var port in Inputs)
        {
            graph.Field(GraphFields.Input, ValueInfo(port));
        }
        foreach (var port in Outputs)
        {
            graph.Field(GraphFields.Output, ValueInfo(port));
        }
    });

    private Protowire Node(Op op) => Protowire.Message(node =>
    {
        // **入と出は宣言した順に並べる。** ONNX は位置で意味が決まる。
        foreach (var input in op.Inputs)
        {
            node.Field(NodeFields.Input, input);
        }
        foreach (var output in op.Outputs)
        {
            node.Field(NodeFields.Output, output);
        }
        node.Field(NodeFields.Name, op.Outputs.Count > 0 ? op.Outputs[0] : op.Type);
        node.Field(NodeFields.OpType, op.Type);
        foreach (var attribute in op.Attributes)
        {
            node.Field(NodeFields.Attribute, Attribute(attribute));
        }
    });

    private static Protowire Attribute(Attr attribute) => Protowire.Message(made =>
    {
        made.Field(AttributeFields.Name, attribute.Name);
        switch (attribute)
        {
            case Attr.Int(_, var value):
                made.Field(AttributeFields.Type, (int)AttributeType.Int);
                made.Field(AttributeFields.I, (ulong)value);
                break;

            case Attr.Ints(_, var values):
                made.Field(AttributeFields.Type, (int)AttributeType.Ints);
                // packed で書く。ONNX は packed も繰り返しも受ける。
                var packed = new List<byte>();
                foreach (var one in values)
                {
                    Protowire.Put(packed, (ulong)one);
                }
                made.Field(AttributeFields.Ints, [.. packed]);
                break;

            case Attr.Float(_, var value):
                made.Field(AttributeFields.Type, (int)AttributeType.Float);
                made.Field(AttributeFields.F, value);
                break;

            case Attr.Text(_, var value):
                made.Field(AttributeFields.Type, (int)AttributeType.String);
                made.Field(AttributeFields.S, Encoding.UTF8.GetBytes(value));
                break;
        }
    });

    private static Protowire Tensor(Initializer weight) => Protowire.Message(made =>
    {
        foreach (var size in weight.Dims)
        {
            made.Field(TensorFields.Dims, size);
        }
        made.Field(TensorFields.DataType, (int)weight.Type);
        made.Field(TensorFields.Name, weight.Name);
        made.Field(TensorFields.RawData, weight.Raw);
    });

    private static Protowire ValueInfo(Port port) => Protowire.Message(made =>
    {
        made.Field(ValueInfoFields.Name, port.Name);
        made.Field(ValueInfoFields.Type, Protowire.Message(type =>
            type.Field(TypeFields.TensorType, Protowire.Message(tensor =>
            {
                tensor.Field(TensorTypeFields.ElemType, (int)port.Type);
                tensor.Field(TensorTypeFields.Shape, Protowire.Message(shape =>
                {
                    foreach (var extent in port.Shape)
                    {
                        shape.Field(ShapeFields.Dim, Protowire.Message(dim =>
                        {
                            if (extent.Name is { } label)
                            {
                                dim.Field(DimensionFields.Param, label);
                            }
                            else
                            {
                                dim.Field(DimensionFields.Value, extent.Fixed);
                            }
                        }));
                    }
                }));
            }))));
    });
}
